cmd/generic-handlers.go

```diff
@@ -469,3 +475,75 @@ func (h pathValidityHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	}
 	h.handler.ServeHTTP(w, r)
 }
+
+// logginHandler logs all HTTP requests and responses
+// and write the log to a io.Writer object.
+type loggingHandler struct {
+	writer    io.Writer
+	writeLock *sync.Mutex
+	handler   http.Handler
+}
+
+func (h loggingHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	const timeFormat = "2006-01-02 15:04:05 -0700"
+	const recordLen = 32 * 1024
+
+	// Generate short random request ID
+	reqID := fmt.Sprintf("%X", time.Now().UnixNano()%1e8)
+
+	// Setup a http request body recorder
+	reqBodyRecorder := &recordReader{Reader: r.Body, recLen: recordLen}
+	r.Body = ioutil.NopCloser(reqBodyRecorder)
+	// Setup a http response body recorder
+	respBodyRecorder := &recordResponseWriter{ResponseWriter: w, recLen: recordLen}
+
+	h.handler.ServeHTTP(respBodyRecorder, r)
+
+	// Build request log and write it to log file
+	reqStatus := fmt.Sprintf("%s %s", r.Method, r.URL.Path)
+	if r.URL.RawQuery != "" {
+		reqStatus += "?" + r.URL.RawQuery
+	}
+	reqStatus += "\n"
+
+	reqHeaders := fmt.Sprintf("Host: %s\n", r.Host)
+	for k, v := range r.Header {
+		reqHeaders += fmt.Sprintf("%s: %s\n", k, v[0])
+	}
+
+	logText := fmt.Sprintf("[%s] [%s] >> %s%s\n%s\n\n", reqID, time.Now().Format(timeFormat),
+		reqStatus,
+		reqHeaders,
+		string(reqBodyRecorder.Data()))
+
+	h.writeLock.Lock()
+	h.writer.Write([]byte(logText))
+	h.writeLock.Unlock()
+
+	// Build response log and write it to log file
+	logText = fmt.Sprintf("[%s] [%s] << %s\n\n", reqID, time.Now().Format(timeFormat),
+		string(respBodyRecorder.Data()))
+	h.writeLock.Lock()
+	h.writer.Write([]byte(logText))
+	h.writeLock.Unlock()
```

Comment:
you can use one lock/unlock block and inside use fmt.Fprintf() to log both req/resp (instead of fmt.Sprintf) to be more mem efficient